### PR TITLE
feat(firebase_crashlytics): v1 rework

### PR DESCRIPTION
**Pull request by Ehesp**
## Description

Builds upon https://github.com/FirebaseExtended/flutterfire/pull/2288 to bring the Crashlytics upgrade to a federated setup.

- [x] Creates a `firebase_crashlytics_platform_interface` package
- [x] Migrates the `firebase_crashlytics` package to integrate with the federated interface setup.

## Changes

The following changes have been made on-top of the original PR:

- [x] Update code docs (typos, more in-depth explanations).
- The `crash` method should crash the app natively:
  - [x] Dart calls method channel.
  - [x] Android Crash 
  - [x] iOS Crash 
- Adds support for `checkForUnsentReports`:
  - [x] Dart
  - [x] Native
- Adds support for `deleteUnsentReports`:
  - [x] Dart
  - [x] Native
- Adds support for `didCrashOnPreviousExecution`:
  - [x] Dart
  - [x] Native
- Adds support for `sendUnsentReports`:
  - [x] Dart
  - [x] Native
- Adds support for `setCrashlyticsCollectionEnabled`:
  - [x] Dart
  - [x] Native
- [x] The `setCustomKey` method now only accepts valid types (int, num, String, bool).

## Local testing

image

## Related Issues

See base PR.

### CODE REVIEW

File: packages/firebase_crashlytics/firebase_crashlytics/lib/src/firebase_crashlytics.dart [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  ///
+  /// Accepts a maximum of 64 key/value pairs. New keys beyond that limit are
+  /// ignored. Keys or values that exceed 1024 characters are truncated.
+  ///
+  /// The value can only be a type [int], [num], [String] or [bool].
+  Future<void> setCustomKey(String key, dynamic value) async {
+    assert(key != null);
+    assert(value != null);
+    assert(value is int || value is num || value is String || value is bool);
+    if (_shouldReportErrors) return;
```
Comment (axel-op) [inline packages/firebase_crashlytics/firebase_crashlytics/lib/src/firebase_crashlytics.dart:199]:
```suggestion
    if (!_shouldReportErrors) return;
```

File: packages/firebase_crashlytics/firebase_crashlytics/lib/src/firebase_crashlytics.dart [hunk 2]
```diff
@@ -24,9 +27,13 @@ class FirebaseCrashlytics extends FirebasePluginPlatform {
...
-    return FirebaseCrashlytics._(
-      app: Firebase.app(),
-    );
+    if (_instance == null) {
+      _instance = FirebaseCrashlytics._(
+        app: Firebase.app(),
+      );
+    }
+
+    return _instance;
```
Comment (axel-op) [inline packages/firebase_crashlytics/firebase_crashlytics/lib/src/firebase_crashlytics.dart:36]:
```suggestion
    return _instance ??= FirebaseCrashlytics(app: Firebase.app());
```

File: packages/firebase_crashlytics/firebase_crashlytics/README.md [hunk 3]
```diff
@@ -91,13 +122,17 @@ void main() {
...
+
 ```dart
-runZoned<Future<void>>(() async {
+runZonedGuarded<Future<void>>(
+  () async {
     // ...
+  },
+  Crashlytics.instance.recordError,
+);
   }, onError: Crashlytics.instance.recordError);
```
Comment (quentinleguennec) [inline packages/firebase_crashlytics/firebase_crashlytics/README.md:136]: I think line 136 should be remove.

File: packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/method_channel/method_channel_crashlytics.dart [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+  @override
+  Future<void> recordError({
+    String exception,
+    String context,
+    String information,
+    List<Map<String, String>> stackTraceElements,
+  }) {
+    return channel
+        .invokeMethod<void>('Crashlytics#recordError', <String, dynamic>{
```
Comment (keisuke-kiriyama) [inline packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/method_channel/method_channel_crashlytics.dart:64]: Is this part correct?
I think `Crashlytics#onError` is correct.
I'm sorry if I misunderstand.
 https://github.com/invertase/flutterfire/blob/24916c4b8d68fefe7c6e372902f0205f7dfafa7b/packages/firebase_crashlytics/firebase_crashlytics/darwin/Classes/FirebaseCrashlyticsPlugin.m#L38

**Comment by talent-apps** [comment]
Please make sure the new crashlytics works properly with obfuscated dart code.
As of now, no stack traces are seen on crash reports making the plugin useless and developers practically blind for crashes in production.
See: #2644

File: packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/method_channel/method_channel_crashlytics.dart [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      'information': information,
+      'stackTraceElements': stackTraceElements ?? [],
+    }).catchError(catchPlatformException);
+  }
+
+  @override
+  Future<void> log(String message) {
+    return channel.invokeMethod<void>('Crashlytics#log', <String, dynamic>{
+      'message': message,
+    });
```
Comment (axel-op) [inline packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/method_channel/method_channel_crashlytics.dart:78]:
```suggestion
    }).catchError(catchPlatformException);
```

File: packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/platform_interface/platform_interface_crashlytics.dart [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+  static final Object _token = Object();
+
+  /// Returns the [FirebaseApp] for the current instance.
+  FirebaseApp get app {
+    if (appInstance == null) {
+      return Firebase.app();
+    }
+
+    return appInstance;
```
Comment (axel-op) [inline packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/platform_interface/platform_interface_crashlytics.dart:32]:
```suggestion
    return appInstance ?? Firebase.app();
```

File: packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/platform_interface/platform_interface_crashlytics.dart [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+  /// The current default [FirebaseFirestorePlatform] instance.
+  ///
+  /// It will always default to [MethodChannelFirebaseFirestore]
+  /// if no other implementation was provided.
+  static FirebaseCrashlyticsPlatform get instance {
+    if (_instance == null) {
+      _instance = MethodChannelFirebaseCrashlytics(app: Firebase.app());
+    }
+    return _instance;
```
Comment (axel-op) [inline packages/firebase_crashlytics/firebase_crashlytics_platform_interface/lib/src/platform_interface/platform_interface_crashlytics.dart:45]:
```suggestion
    return _instance ??= MethodChannelFirebaseCrashlytics(app: Firebase.app());
```

File: packages/firebase_crashlytics/firebase_crashlytics/lib/src/firebase_crashlytics.dart [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+  // Cached and lazily loaded instance of [FirebaseCrashlyticsPlatform] to avoid
+  // creating a [MethodChannelFirebaseCrashlytics] when not needed or creating an
+  // instance with the default app before a user specifies an app.
+  FirebaseCrashlyticsPlatform _delegatePackingProperty;
+
+  FirebaseCrashlyticsPlatform get _delegate {
+    if (_delegatePackingProperty == null) {
+      _delegatePackingProperty = FirebaseCrashlyticsPlatform.instance;
+    }
+    return _delegatePackingProperty;
```
Comment (axel-op) [inline packages/firebase_crashlytics/firebase_crashlytics/lib/src/firebase_crashlytics.dart:19]:
```suggestion
    return _delegatePackingProperty ??= FirebaseCrashlyticsPlatform.instance;
```

File: packages/firebase_crashlytics/firebase_crashlytics/README.md [hunk 9]
```diff
@@ -12,43 +12,72 @@ For Flutter plugins for other Firebase products, see [README.md](https://github.
 
 To use the `firebase_crashlytics` plugin, follow the [plugin installation instructions](https://pub.dev/packages/firebase_crashlytics#pub-pkg-tab-installing).
 
+The following instructions are from [the official installation page](https://firebase.google.com/docs/crashlytics/get-started-new-sdk).
+
```
Comment (HofmannZ) [inline packages/firebase_crashlytics/firebase_crashlytics/README.md:16]: For iOS, you might want to mention to skip step 2.

File: packages/firebase_crashlytics/firebase_crashlytics/ios/firebase_crashlytics.podspec [hunk 10]
```diff
@@ -1,29 +1,43 @@
...
-  s.dependency 'Firebase/Analytics'
+
+  s.dependency 'firebase_core'
+  s.dependency 'Firebase/CoreOnly', "~> #{firebase_sdk_version}"
+  s.dependency 'Firebase/Crashlytics', "~> #{firebase_sdk_version}"
   
-  s.pod_target_xcconfig = { 'GCC_PREPROCESSOR_DEFINITIONS' => "LIBRARY_VERSION=\\@\\\"#{libraryVersion}\\\" LIBRARY_NAME=\\@\\\"flutter-fire-cls\\\"" }
+  s.static_framework = true
+  s.pod_target_xcconfig = { 
+    'GCC_PREPROCESSOR_DEFINITIONS' => "LIBRARY_VERSION=\\@\\\"#{library_ersion}\\\" LIBRARY_NAME=\\@\\\"flutter-fire-cls\\\"",
```
Comment (HofmannZ) [inline packages/firebase_crashlytics/firebase_crashlytics/ios/firebase_crashlytics.podspec:40]:
```suggestion
    'GCC_PREPROCESSOR_DEFINITIONS' => "LIBRARY_VERSION=\\@\\\"#{library_version}\\\" LIBRARY_NAME=\\@\\\"flutter-fire-cls\\\"",
```

Comment (Salakar) [reply packages/firebase_crashlytics/firebase_crashlytics/ios/firebase_crashlytics.podspec]: Already fixed locally, have pushed :P thanks!

File: packages/firebase_crashlytics/firebase_crashlytics/README.md (hunk 9 above)
Comment (Salakar) [reply packages/firebase_crashlytics/firebase_crashlytics/README.md]: Have pushed up the docs, this was the old readme. Would you mind reviewing docs in the `docs/crashlytics` folder? Thanks :)

File: docs/crashlytics/overview.mdx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+apply plugin: 'com.google.firebase.crashlytics'
+```
+
+#### b. iOS
+
+1. From Xcode select `Runner` from the project navigation.
+2. Select the `Build Phases` tab, then click `+ > New Run Script Phase`.
+3. Add `${PODS_ROOT}/FirebaseCrashlytics/run` to the `Type a script...` text box.
+4. You can also provide your app's built `Info.plist` location to the build phase's Input Files field:
+For example: `$(BUILT_PRODUCTS_DIR)/$(INFOPLIST_PATH)`
```
Comment (HofmannZ) [inline docs/crashlytics/overview.mdx:70]: Nitpicking: This is optional, right? Should we mention that?

File: docs/crashlytics/overview.mdx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  implementation 'com.google.firebase:firebase-crashlytics:17.2.1'
+}
+
+apply plugin: 'com.google.firebase.crashlytics'
+```
+
+#### b. iOS
+
+1. From Xcode select `Runner` from the project navigation.
+2. Select the `Build Phases` tab, then click `+ > New Run Script Phase`.
```
Comment (HofmannZ) [inline docs/crashlytics/overview.mdx:67]: Nitpicking: By default, XCode adds the build phase to the end of the list. Once you build the Flutter app for the first time, it adds some extra build phases for managing the CocaoPods. Does that play nice with that?

**Comment by IchordeDionysos** [comment]
Whats's the status of this PR?
It seems quite far in the process, even Docs are already in place! 👍

**Comment by Salakar** [comment]
> Whats's the status of this PR?
> It seems quite far in the process, even Docs are already in place! 👍

I believe as of now, it's ready to go, pending review of course.

**Comment by Salakar** [comment]
Pending CI I'll get this on the main repo as a PR and then we can do final review there and any final cleanup bits.

**Pull request merged**